### PR TITLE
[WIP] add entanglement entropy

### PR DESCRIPTION
**Pull request by PhilipVinc**
needs tests, but please let me know if that code style is ok

### CODE REVIEW

File: src/metrics.jl [hunk 1]
```diff
@@ -236,3 +236,45 @@ function avg_gate_fidelity(x::T, y::T) where T <: Union{PauliTransferMatrix{B, B
...
 end
+
+"""
+    entanglement_entropy(state, partition, [entropy_fun=entropy_vn])
+
+Computes the entanglement entropy of `state` between the list of sites `partition`
+and the rest of the system. The state must be defined in a composite basis.
+
+By default the entropy is computed according to Von-Neumann entropy, but a different
+function can be provided (for example to compute the entanglement-reny entropy).
```
Comment (david-pl) [inline src/metrics.jl]: Typo here: it's *renyi*

File: src/metrics.jl [hunk 2]
```diff
@@ -236,3 +236,45 @@ function avg_gate_fidelity(x::T, y::T) where T <: Union{PauliTransferMatrix{B, B
...
+"""
+function entanglement_entropy(psi::Ket{B}, partition, entropy_fun=entropy_vn) where B<:CompositeBasis
```
Comment (david-pl) [inline src/metrics.jl]: The `if partition isa Number` can be handled more nicely by a dispatch. Add type constraints here `partition::Vector{Int}`. 
Then, another method below like this `entanglement_entropy(state, partition::Number, args...) = entanglement_entropy(state, Int[partition], args...)` deals with number partitions.

File: src/metrics.jl [hunk 3]
```diff
@@ -236,3 +236,45 @@ function avg_gate_fidelity(x::T, y::T) where T <: Union{PauliTransferMatrix{B, B
...
+    end
+
+    # check that sites are within the range
+    @assert all(partition .<= length(psi.basis.bases))
+
+    rho = ptrace(psi, partition)
+    return entropy_fun(rho)
+end
+
+function entanglement_entropy(rho::DenseOperator{B,B}, partition, entropy_fun=entropy_vn) where {B<:CompositeBasis}
```
Comment (david-pl) [inline src/metrics.jl]: As above: `partition::Vector{Int}`

File: test/test_metrics.jl [hunk 4]
```diff
@@ -90,6 +90,13 @@ rho = rho ⊗ dm(g)
 @test logarithmic_negativity(rho, 1) == logarithmic_negativity(rho, 2) ≈ 1.0
 @test isapprox(logarithmic_negativity(rho, 3), 0.0, atol=1e-15)
 
+# Entanglement Entropy
+b1 = SpinBasis(1//2)
+psi = 1/sqrt(2)*(spinup(b1)⊗spindown(b1) + spindown(b1)⊗spinup(b1))
+rho_ent = dm(psi)
+rho_mix = DenseOperator(rho_ent.basis_l, diagm(ComplexF64[1.0,1.0,1.0,1.0]))
+@test entanglement_entropy(rho_mix, 1) ≈ 0.0
+
```
Comment (david-pl) [inline test/test_metrics.jl]: You don't actually test `entanglement_entropy(rho_ent)` here. Is that on purpose?

**Comment by david-pl** [comment]
I left some suggestions in a review, but in principle everything looks good to me. Let me know when you're done, then I'll merge.

File: test/test_metrics.jl (hunk 4 above)
Comment (PhilipVinc) [reply test/test_metrics.jl]: ops, my bad.

**Comment by PhilipVinc** [comment]
About the `partition isa Number`. 
I dislike restricting types too much. For example, where I to restrict this to Vector{Int} (or `AbstractVector{<:Integer}`) while `1:5` would work, the tuple `(1,2,3,4,5)` would not work.

The good constraint is that it should be an iterable...

**Comment by PhilipVinc** [comment]
After discussion with a few colleagues, I realised I should also add to the doctoring the note that this computes the entanglement-entropy for Kets, and the operator-space entanglement entropy for Operators. 

Which explains why `entanglement_entropy(ket) = 0.5 * entanglement_entropy(dm(ket))

**Comment by david-pl** [comment]
Alright, the dispatch was just a suggestion. Let me know when to merge.

**Comment by PhilipVinc** [comment]
If tests pass, this should be ready to merge.

I have added the type constraint like you proposed.
It would be the scope of another PR to relax all those `::Vector{Int}` type constraints. 

It turns out that they can all be gotten rid of (thanks to the fact that numbers are iterables of length 1), except that deep in `ptrace->check_ptrace_arguments->check_indices` that method should be rewritten.

**Comment by PhilipVinc** [comment]
Tests are passing.
The change in coverage is not my fault (I think it's due to how Julia recently started reporting coverage differently...)

I also bumped the version, so that you can quickly tag a new release, which I'd appreciate.

**Pull request merged**